### ISSUE TITLE
Setting `checkout.workers` causes checkouts of multiple files to fail

### ISSUE THREAD

**Issue by Vynce** [git-for-windows/git#4331]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.39.1.windows.1
cpu: x86_64
built from commit: b03dafd9c26b06c92d509a07ab01b01e6d0d85ee
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

Windows 11 22H2 64-bit. Also tried Windows 10 22H2.
```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.22621.1265]
Microsoft Windows [Version 10.0.19045.2673]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

Also tried all the defaults.
```
Editor Option: Nano
Custom Editor Path:
Default Branch Option:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: WinSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Rebase
Use Credential Manager: Enabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Enabled
Enable FSMonitor: Disabled
```

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Windows Terminal 1.17. Also tried Git Bash.

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

Also see attached demo repo: [checkout_workers_test.zip](https://github.com/git-for-windows/git/files/10914970/checkout_workers_test.zip)
```
echo >> a.txt; echo >> b.txt; echo >> c.txt
git -c checkout.workers=0 checkout a.txt b.txt c.txt
```
 - What did you expect to occur after running these commands?

```
Updated 3 paths from the index
```

 - What actually happened instead?

```
error: unable to stat just-written file 'b.txt': No such file or directory
error: unable to stat just-written file 'c.txt': No such file or directory
Updated 1 path from the index
```

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

Minimal repo demonstrating the issue is attached: [checkout_workers_test.zip](https://github.com/git-for-windows/git/files/10914970/checkout_workers_test.zip)

**Comment by dscho** [comment]
This smells like a hyperactive anti-virus software. Files that have been just written should be there.

On the other hand, it could be the FSCache. Do run with `git -c core.fscache=false [...]` and see whether that works around the issue.

**Comment by Vynce** [comment]
Thanks for the fast response!

It works correctly with FSCache disabled (`core.fscache=false`). Seems like  `core.fscache` and `checkout.workers` are mutually exclusive. Not a huge issue, but was not obvious at all 😊.

FWIW, I also tested in a fresh Windows 10 VM. Windows Defender was enabled there, but there were no other security tools installed.

**Comment by dscho** [comment]
> Seems like `core.fscache` and `checkout.workers` are mutually exclusive.

Yes, and that's a bug. We most likely need to insert a `flush_fscache()` call in some strategic area, probably appending it after https://github.com/git-for-windows/git/blob/v2.41.0-rc2.windows.1/parallel-checkout.c#L81 would work.

@Vynce as you can probably tell by the _insane_ delay in my answer, I am so under water that I cannot even test this suggestion. Maybe you can? To get started:
1. install [Git for Windows' SDK](https://gitforwindows.org/#download-sdk),
2. `sdk cd git`,
3. edit `parallel-checkout.c`
4. build Git via `make -j$(nproc)`
5. test in-place via `./git --exec-path="$PWD" -C <directory> <command>`?
6. open a PR?
